### PR TITLE
chore: separate templates

### PR DESCRIPTION
**Pull request by cam-garrison**
## Description

RHOAIENG-4604

This PR separates out the templates that are embedded and applied as part of Feature manifests. The goal here is to make it easier to find the relevant templates for the component they are used, so we move templates from a central directory under `pkg/feature/templates` into their own `/templates` directories, IE `components/kserve/templates` now stores the templates used in deploying kserve with service mesh and `controllers/dscinitialization/templates` holds the templates used in deploying service mesh.    

This PR also renames our `.tmpl` files to `.tmpl.yaml` so that IDEs will recognize and format these files accordingly. 

Note that this changes the behavior, so that when supplying a Feature with Manifests(paths) to files within an embedded file system, one has to first set the ManifestSource() to the embeddedFS. 

## How Has This Been Tested?

Tested on CRC with image `quay.io/cgarriso/opendatahub-operator:dev-separate-templates` and ensured that kserve + dsci features still applied successfully. 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/dscinitialization/servicemesh_setup.go
Comment (zdtsw) [inline controllers/dscinitialization/servicemesh_setup.go:83]: maybe i missed something, just try to understand why some use dir, but others list all files:
say under "metricsDir" 2 files,  under "meshDir" one file
but line 60 points to that template yaml , but here is use the folder.
should line 60 not be 
`path.Join(meshDir),`
than 
`path.Join(meshDir, "create-smcp.tmpl.yaml"),`

File: controllers/dscinitialization/servicemesh_setup.go
Comment (zdtsw) [inline controllers/dscinitialization/servicemesh_setup.go:104]:
```suggestion
                path.Join(authorinoDir, "base"),
				path.Join(authorinoDir, "auth-smm.tmpl.yaml"),				
```

File: controllers/dscinitialization/feature_templates.go
Comment (zdtsw) [inline controllers/dscinitialization/feature_templates.go:14]: can we rename this folder to "servicemesh" just match whats named under kserve?

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
I have some suggestions for the structure of the code, but I am not also quite sure about the directories. If we are moving templates around maybe we can also improve the folders.

In particular for KServe I think it would be better to group templates by capability/infra, so that would look like:

```
├── servicemesh
│   ├── activator-envoyfilter.tmpl.yaml
│   ├── envoy-oauth-temp-fix.tmpl.yaml
│   ├── grpc-envoyfilter-temp-fix.tmpl.yaml
│   ├── kserve-predictor-authorizationpolicy.tmpl.yaml
│   ├── routing
│   │   ├── istio-ingress-gateway.tmpl.yaml
│   │   ├── istio-local-gateway.yaml
│   │   └── local-gateway-svc.tmpl.yaml
│   └── z-migrations
│       └── kserve-predictor-authorizationpolicy.patch.tmpl.yaml
├── serving-install
│   ├── knative-serving.tmpl.yaml
│   └── service-mesh-subscription.tmpl.yaml
├── serving-istio-gateways
└── serving-net-istio-secret-filtering.patch.tmpl.yaml
```

With regards to `fs.FS` and kyaml's `filesys.FileSystem` or in general support for Kustomize. In a retrospect, I think making it available right now might be a bit premature. We do not have the whole thing in parity with `deploy.go` (especially in terms of feature create/update/remove path from the reconcile perspective). At the time we start looking at how to unify the approach we can bring this code back. Perhaps best for now is to make a separate PR first removing this part (so we can restore easier at later stage) and then get this one in.

File: controllers/dscinitialization/feature_templates.go
Comment (cam-garrison) [reply controllers/dscinitialization/feature_templates.go]: Agreed, done 
1099d0a

File: controllers/dscinitialization/servicemesh_setup.go
Comment (cam-garrison) [reply controllers/dscinitialization/servicemesh_setup.go]: done here, thx 
1099d0a

File: controllers/dscinitialization/servicemesh_setup.go
Comment (cam-garrison) [reply controllers/dscinitialization/servicemesh_setup.go]: +1 
1099d0a

File: components/kserve/servicemesh_setup.go
Comment (cam-garrison) [reply components/kserve/servicemesh_setup.go]: Agree for checking basepath there - [a945320](https://github.com/opendatahub-io/opendatahub-operator/pull/932/commits/a9453203b341c09596f8aafe02e8bf18c8a2c388)

But I think checking for `.patch.` or .`tmp.` would be redundant since `.patch.` contains `.patch`

File: tests/integration/features/manifests_int_test.go
Comment (cam-garrison) [reply tests/integration/features/manifests_int_test.go:43]: Good catch, re added
3a06b3f

File: components/kserve/servicemesh_setup.go
Comment (bartoszmajsak) [reply components/kserve/servicemesh_setup.go]: My thinking here is that `.tmp.` is more explicit in terms of format.

File: components/kserve/serverless_setup.go
Comment (bartoszmajsak) [inline components/kserve/serverless_setup.go:38]: Worth making it part of `Templates` struct? I also wonder if this is a good name. After all it's not only templates we are keeping there.

File: components/kserve/feature_templates.go
Comment (bartoszmajsak) [inline components/kserve/feature_templates.go:22]: I am not sure if `.Files` is the right name. `.Location` or even `.Source`. WDYT?

File: pkg/feature/builder.go
Comment (bartoszmajsak) [inline pkg/feature/builder.go:213]: nit, but maybe worth moving it higher up in the file so it's aligned with the chain.

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Just a few nits to consider.

Please resolve merge conflicts, as kustomize stuff is gone now :)

File: components/kserve/feature_templates.go
Comment (cam-garrison) [reply components/kserve/feature_templates.go:22]: I think `.Source` could be best :+1:

File: components/kserve/servicemesh_setup.go
Comment (cam-garrison) [reply components/kserve/servicemesh_setup.go]: Ah, I understand now, we can force our naming convention with this. Will do.

File: pkg/feature/builder.go
Comment (cam-garrison) [reply pkg/feature/builder.go]: done, moved both up

File: components/kserve/serverless_setup.go
Comment (cam-garrison) [reply components/kserve/serverless_setup.go]: renamed the struct and included `Resources.BaseDir` on top of the const `baseDir` to be used (so the const can still be used in creating the fields for the subpaths

**Comment by cam-garrison** [comment]
@bartoszmajsak re-requested review as I've re-done the kserve smoke test as discussed and it worked

**Review by zdtsw — APPROVED** [review]

**Review by bartoszmajsak — APPROVED** [review]
LGTM

I intentionally put it on hold as I would like @israel-hdez to sign off on this one before we merge it.

**Review by israel-hdez — APPROVED** [review]
> I intentionally put it on hold as I would like @israel-hdez to sign off on this one before we merge it.

You have my :+1:. All works fine for me after the changes.

I'm dropping two non-blocking comments.

File: controllers/dscinitialization/resources/metrics-collection/envoy-metrics-collection.tmpl.yaml
Comment (bartoszmajsak) [reply controllers/dscinitialization/resources/metrics-collection/envoy-metrics-collection.tmpl.yaml:1]: Good point, we will take care of it when moving stuff around, though the feature is already in DSCI part at the moment :)

File: controllers/dscinitialization/servicemesh_setup.go
Comment (bartoszmajsak) [reply controllers/dscinitialization/servicemesh_setup.go:174]: Good catch, I think the order should be preserved @cam-garrison

Comment (bartoszmajsak) [reply controllers/dscinitialization/servicemesh_setup.go:174]: > If we successfully migrate from patching the Authorino deployment at PostConditions, to using Labels via the Authorino operator CRD, I think it makes more sense the previous ordering. Isn't it?

I will create JIRA for this (last famous words)

Comment (cam-garrison) [reply controllers/dscinitialization/servicemesh_setup.go]: thanks for catching this! reverted to prior order here 
d4731cc

Comment (bartoszmajsak) [reply controllers/dscinitialization/servicemesh_setup.go]: https://issues.redhat.com/browse/RHOAIENG-5494

**Review by bartoszmajsak — APPROVED** [review]
LGTM

**Pull request merged**
